org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/text/correction/QuickAssistProcessor.java

```diff
@@ -622,4 +634,152 @@ private static Block getBlockBodyForLambda(Expression bodyExpr, ITypeBinding ret
 		blockBody.statements().add(statementInBlockBody);
 		return blockBody;
 	}
+
+	public static boolean getCatchClauseToThrowsProposals(IInvocationContext context, ASTNode node, Collection<CUCorrectionProposal> resultingCollections) {
+		CatchClause catchClause = (CatchClause) ASTResolving.findAncestor(node, ASTNode.CATCH_CLAUSE);
+		if (catchClause == null) {
+			return false;
+		}
+
+		Statement statement = ASTResolving.findParentStatement(node);
+		if (statement != catchClause.getParent() && statement != catchClause.getBody()) {
+			return false; // selection is in a statement inside the body
+		}
+
+		Type type = catchClause.getException().getType();
+		if (!type.isSimpleType() && !type.isUnionType() && !type.isNameQualifiedType()) {
+			return false;
+		}
+
+		BodyDeclaration bodyDeclaration = ASTResolving.findParentBodyDeclaration(catchClause);
+		if (!(bodyDeclaration instanceof MethodDeclaration) && !(bodyDeclaration instanceof Initializer)) {
+			return false;
+		}
+
+		if (resultingCollections == null) {

```

Comment:
move that block at the beginning of the method, no need to compute anything if a simple check fails

Reply:
Updated